### PR TITLE
Add bfit bolditalic tex cmd

### LINKED ISSUE #19393: \bf\it in mathtext
### Problem
`\bf` and `\it` are both supported in mathtext but if used together, only the inner style is applied.

### Proposed Solution

Apply both styles if both `\bf` and `\it` are used together. Alternatively, a new `\bfit` command could be introduced which enables text to be set in bold and italic.

### Additional context and prior art

This would be very useful when setting the experiment label in figures for the ATLAS experiment, which requires the name of the experiment in bold-italic followed by text describing the status of that figure in ordinary text. Additional text below this label communicates various experimental conditions that are often useful in interpreting the figure. In the absence of rich text functionality, mathtext (or `usetex`) is the only way to do this.

I currently handle this in [beojan/atlas-mpl](github.com/beojan/atlas-mpl) by using `usetex`. However, this requires the user to have a full LaTeX installation available, along with the TeX Gyre fonts in that installation in order to meet the font family requirements of the experiment. 

@andrzejnovak handles the same issue in [scikit-hep/mplhep](github.com/scikit-hep/mplhep) by setting the various parts of the label separately, heuristically calculating the offsets needed. This is suboptimal compared to using the text layout algorithms built into the Matplotlib text engine.

### PR DESCRIPTION
**Pull request by devRD**
## PR Summary
Fixes #19393
Adds support for bfit command in mathtext for bold italic

## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by anntzer** [comment]
Is that only implemented for stix?  Do dejavu & cm also need their own thing?

**Comment by devRD** [comment]
Looks like dejavu and cm require their own mappings.

File: lib/matplotlib/tests/test_mathtext.py [hunk 1]
```diff
@@ -151,6 +151,7 @@
     (['mathrm'], all),
     (['mathbf'], all),
     (['mathit'], all),
+    (['mathbfit'], all),
```
Comment (rcomer) [inline lib/matplotlib/tests/test_mathtext.py:154]: The comment just above this list says “Add new tests at the end.”  Looking at the failing tests themselves, it looks like the images are numbered according to the index in the list.  So I think you just need to move your new entry to the end of the list.

**Comment by oscargus** [comment]
This is probably a good location to document it:
https://matplotlib.org/stable/tutorials/text/mathtext.html#fonts

**Comment by oscargus** [comment]
Thanks! Would it also make sense to add it here? It seems to be working for stix right? (And then also for deja vu, but there is no special section for those I guess and maybe out of scope to add one.)
image

Also, for both entries here, it says "bold italic"?
image

**Comment by devRD** [comment]
I am not exactly sure how to add an alias for the `\mathbfit` command but I'll look into it. If you have any references I would appreciate it.  

And, I believe the `\mathbf` command stating bold italic is incorrect as it just provides the boldface font. I can update both changes in a new commit.

**Comment by oscargus** [comment]
> I am not exactly sure how to add an alias for the `\mathbfit` command but I'll look into it.

I just checked the code and I do not think that there is an easy way to do that...

Probably one will have to add a special command for this, both wrt parsing and a callback.

**Comment by anntzer** [comment]
@oscargus Per https://github.com/matplotlib/matplotlib/issues/19393#issuecomment-1013873265, for consistency with latex, I think we should actually make `\mathfoo{\mathbar{...}}` *not* stack the styles, but just overwrite one another, and introducing `\mathbfit` for the stacked version seems correct.

**Comment by devRD** [comment]
Hi, @rcomer @oscargus @anntzer

Could you please review this patch and let me know if there are any further modifications required?

**Comment by rcomer** [comment]
I'm afraid I don't have the expertise to review this one.

**Comment by oscargus** [comment]
I think that the tutorial should be updated a bit further as discussed above (at least add it to the Stix section).

I'm also wondering if `#mathtext.bfit:sans` is the correct default rcParam, but I do not know exactly how that works. (I'd guess something like `#mathtext.bfit:sans:bold:italic`, but not sure these can be stacked.)

File: lib/matplotlib/mpl-data/matplotlibrc [hunk 2]
```diff
@@ -340,6 +340,7 @@
 ## settings which map a TeX font name to a fontconfig font pattern.  (These
 ## settings are not used for other font sets.)
 #mathtext.bf:  sans:bold
+#mathtext.bfit:sans
```
Comment (oscargus) [inline lib/matplotlib/mpl-data/matplotlibrc:343]:
```suggestion
#mathtext.bfit: sans:italic:bold
```
This is what I think is a more correct default.

File: galleries/tutorials/text/mathtext.py [hunk 3]
```diff
@@ -215,6 +215,7 @@
     ``\mathfrak{Fraktur}``           :math-stix:`\mathfrak{Fraktur}`
     ``\mathsf{sansserif}``           :math-stix:`\mathsf{sansserif}`
     ``\mathrm{\mathsf{sansserif}}``  :math-stix:`\mathrm{\mathsf{sansserif}}`
+    ``\mathbfit{blackboard}``        :math-stix:`\mathbfit{blackboard}}`
```
Comment (oscargus) [inline galleries/tutorials/text/mathtext.py:218]:
```suggestion
    ``\mathbfit{bolditalic}``        :math-stix:`\mathbfit{bolditalic}}`
```

File: doc/users/next_whats_new/updated_mathtext.rst [hunk 4]
```diff
@@ -0,0 +1,14 @@
+Bold-italic cmd ``\mathbfit``
+-----------------------------------------------------
+
+Supports use of bold-italic font style in Mathtext using the *\mathbfit{}* command:
```
Comment (oscargus) [inline doc/users/next_whats_new/updated_mathtext.rst:4]:
```suggestion
Supports use of bold-italic font style in mathtext using the ``\mathbfit{}`` command:
```

File: galleries/users_explain/text/mathtext.py [hunk 5]
```diff
@@ -220,6 +220,7 @@
     ``\mathfrak{Fraktur}``           :math-stix:`\mathfrak{Fraktur}`
     ``\mathsf{sansserif}``           :math-stix:`\mathsf{sansserif}`
     ``\mathrm{\mathsf{sansserif}}``  :math-stix:`\mathrm{\mathsf{sansserif}}`
+    ``\mathbfit{bolditalic}``        :math-stix:`\mathbfit{bolditalic}}`
```
Comment (oscargus) [inline galleries/users_explain/text/mathtext.py:223]:
```suggestion
    ``\mathbfit{bolditalic}``        :math-stix:`\mathbfit{bolditalic}`
```

**Review by oscargus — COMMENTED** [review]
Just some minor details, but all-in-all looking good!

**Comment by devRD** [comment]
Hi @oscargus, I have updated the PR with the suggestions. Please let me know if anything else needs to be updated.

**Review by oscargus — APPROVED** [review]
Thanks!

A reminder to whoever may merge this to squash the commits. (@devRD if you know how to squash, please do.)

File: lib/matplotlib/_mathtext.py [hunk 6]
```diff
@@ -1771,7 +1774,8 @@ class _MathStyle(enum.Enum):
 
     _dropsub_symbols = set(r'''\int \oint'''.split())
 
-    _fontnames = set("rm cal it tt sf bf default bb frak scr regular".split())
+    _fontnames = set("""rm cal it tt sf bf bfit
+                     default bb frak scr regular""".split())
```
Comment (anntzer) [inline lib/matplotlib/_mathtext.py:1778]: could be
```
_fontname = set("..."
                "...".split())
```
which seems nicer than a multiline string

File: lib/matplotlib/_mathtext_data.py [hunk 7]
```diff
@@ -1210,6 +1210,13 @@
             (0x2202, 0x2202, 'bf', 0x1d789), # partial differential
             (0x2207, 0x2207, 'bf', 0x1d76f), # \Nabla
             ],
+        'bfit':
+            [
+            (0x0041, 0x005a, 'bfit', 0x1d468), # A-Z
+            (0x0061, 0x007a, 'bfit', 0x1d482), # a-z
+            (0x0393, 0x03a9, 'bfit', 0x1d71E), # \Gamma-\Omega
```
Comment (anntzer) [inline lib/matplotlib/_mathtext_data.py:1217]: lowercase e for consistency

File: lib/matplotlib/tests/test_mathtext.py [hunk 8]
```diff
@@ -167,7 +167,8 @@
     (['mathscr'], [uppercase, lowercase]),
     (['mathsf'], [digits, uppercase, lowercase]),
     (['mathrm', 'mathsf'], [digits, uppercase, lowercase]),
-    (['mathbf', 'mathsf'], [digits, uppercase, lowercase])
+    (['mathbf', 'mathsf'], [digits, uppercase, lowercase]),
+    (['mathbfit'], all)
```
Comment (anntzer) [inline lib/matplotlib/tests/test_mathtext.py:171]: add a trailing quote here

**Review by anntzer — APPROVED** [review]
Extremely minor nits, but otherwise lgtm.

File: lib/matplotlib/tests/test_mathtext.py (hunk 8 above)
Comment (QuLogic) [reply lib/matplotlib/tests/test_mathtext.py:171]: Trailing comma?

Comment (anntzer) [reply lib/matplotlib/tests/test_mathtext.py:171]: yes, sorry.

File: lib/matplotlib/_mathtext.py (hunk 6 above)
Comment (ksunden) [reply lib/matplotlib/_mathtext.py:1778]: In fact, now that we have bumped the line length to 88, this all fits on one line within our guidelines, still

**Comment by devRD** [comment]
Hi all!
Thank you for the reviews. I have made the changes and rebased to the current main.

**Review by anntzer — APPROVED** [review]

**Pull request merged**